### PR TITLE
Store Language Name, Version, and Requirements for `npm`, `pnpm`, and `yarn`

### PR DESCRIPTION
**Pull request by kbukum1**
### What are you trying to accomplish?

This PR adds functionality to detect and include the language name (`Node.js`) and its version for `npm`, `pnpm`, and `yarn` package managers. 

**Why?**
- To accurately identify the runtime environment (`Node.js`) by using `node -v` to fetch the version directly.
- This ensures dependency updates are handled with the correct language version context, improving reliability in JavaScript-based ecosystems.

This change improves Dependabot's ability to work with JavaScript projects by explicitly detecting and documenting Node.js as the runtime language and its version.

---

### Anything you want to highlight for special attention from reviewers?

- The implementation uses the `node -v` command to fetch the Node.js version at runtime. This approach is simple, reliable, and avoids relying on indirect metadata (e.g., `volta` or `engines.node`).
- Test coverage ensures that all three package managers (`npm`, `pnpm`, and `yarn`) work seamlessly with this addition.

---

### How will you know you've accomplished your goal?

- Unit tests validate the language detection and version fetching logic.
- Manually testing against sample `npm`, `pnpm`, and `yarn` projects confirms the correct Node.js version is detected.
- Dependabot updates dependencies without issues when language-specific constraints are considered.

---

### Checklist

- [x] I have run the complete test suite to ensure all tests and linters pass.
- [x] I have thoroughly tested my code changes to ensure they work as expected, including adding additional tests for new functionality.
- [x] I have written clear and descriptive commit messages.
- [x] I have provided a detailed description of the changes in the pull request, including the problem it addresses, how it fixes the problem, and any relevant details about the implementation.
- [x] I have ensured that the code is well-documented and easy to understand.

### CODE REVIEW

**Review by abdulapopoola — APPROVED** [review]

File: npm_and_yarn/lib/dependabot/npm_and_yarn/package_manager.rb [hunk 1]
```diff
@@ -256,6 +269,41 @@ def name_from_engines
       end
     end
 
+    class Language < Ecosystem::VersionManager
+      extend T::Sig
+      NAME = "node"
```
Comment (jonabc) [inline npm_and_yarn/lib/dependabot/npm_and_yarn/package_manager.rb:274]: this is a bit confusing.  node is a runtime, javascript is the language.  What does a language class for `"node"` mean?

Comment (jonabc) [reply npm_and_yarn/lib/dependabot/npm_and_yarn/package_manager.rb:274]: also, what does it mean to have something representing node as an `Ecosystem::VersionManager`? node doesn't manage versions, packages, or anything else.

File: npm_and_yarn/lib/dependabot/npm_and_yarn/helpers.rb [hunk 2]
```diff
@@ -323,6 +323,37 @@ def self.run_npm_command(command, fingerprint: command)
         end
       end
 
+      sig { returns(T.nilable(String)) }
+      def self.node_version
```
Comment (jonabc) [inline npm_and_yarn/lib/dependabot/npm_and_yarn/helpers.rb:327]: It looks like this function only gets called from a single location.  Will it be used more broadly in the future?  If not, is there another reason to add this as a helper and not define it next to where it's used in the package manager, or defined directly on the node language class?

**Review by jonabc — COMMENTED** [review]
it's a little hard to see the functional changes amid all of the added logging.  In the future if you're adding logging to otherwise untouched code, can you bundle it as a separate change/PR?

Re: the new logging - much of the added `Dependabot::Logger.info` output looks like debug information which can get noisy if it's always written to logging.  Would it make sense to use `Dependabot::Logger.debug` to keep noise down in prod logs but still make it easy to toggle on when debugging?

File: npm_and_yarn/lib/dependabot/npm_and_yarn/package_manager.rb (hunk 1 above)
Comment (kbukum1) [reply npm_and_yarn/lib/dependabot/npm_and_yarn/package_manager.rb:274]: @jonabc  
Thank you for the feedback! I appreciate your thoughts, and I’ll address them below.  

Regarding your point about Node.js and JavaScript: While Node.js is technically a runtime for JavaScript, in the context of dependency and manifest files, Node.js often represents the JavaScript environment. For package managers like npm, pnpm, and yarn, we track Node.js as the primary requirement, as it determines the version of JavaScript supported. JavaScript as a language does not have version-specific requirements in the same way that Node.js does, so in practice, Node.js is used to represent the JavaScript environment.  

As for the `VersionManager`, it is designed to handle version-related functionality across ecosystems, whether for package managers or languages. Specifically, it serves the following purposes:  
- Storing raw versions of package managers or languages.  
- Managing version-specific requirements, such as dependencies on a minimum or specific version.  
- Facilitating functionality like unsupported or deprecated versions, which can be applied to both package managers and languages.  

This is why the `VersionManager` applies to both package managers and languages—it consolidates this version management logic in one place. Let me know if there’s any confusion I can clear up or if you’d like me to tweak the approach further.

File: npm_and_yarn/lib/dependabot/npm_and_yarn/helpers.rb (hunk 2 above)
Comment (kbukum1) [reply npm_and_yarn/lib/dependabot/npm_and_yarn/helpers.rb:327]: @jonabc  
While it’s currently used in one place, having it in a helper makes future updates easier if we need to retrieve the Node.js version differently (e.g., from the environment). It also centralizes logic that might be reused, especially for commands dependent on the Node.js version. Generally, we use helpers to manage command execution logic that may change across package managers.

**Comment by kbukum1** [comment]
> it's a little hard to see the functional changes amid all of the added logging. In the future if you're adding logging to otherwise untouched code, can you bundle it as a separate change/PR?
> 
> Re: the new logging - much of the added `Dependabot::Logger.info` output looks like debug information which can get noisy if it's always written to logging. Would it make sense to use `Dependabot::Logger.debug` to keep noise down in prod logs but still make it easy to toggle on when debugging?

@jonabc 
When monitoring metrics, it was challenging to identify which version was being used and what information was captured. This led me to believe that logging this general information properly at the `info` level would be helpful. It allows anyone investigating an issue to clearly see the package manager and language versions involved, potentially replicating the problem with the same versions. For this reason, I felt `info` was more appropriate than `debug`.  

Let me know if you think this could be approached differently!

**Review by jonabc — APPROVED** [review]
I don't have enough practical experience with dependabot-core to know whether my previous comments are valid.  I'm still a little uncomfortable with the change but 🤷 the functional changes seem small and tested :shipit:

**Pull request merged**